lib/api/src/env.rs

```diff
@@ -0,0 +1,202 @@
+use crate::{ExportError, Instance};
+use thiserror::Error;
+
+/// An error while initializing the user supplied host env with the `WasmerEnv` trait.
+#[derive(Error, Debug)]
+#[error("Host env initialization error: {0}")]
+pub enum HostEnvInitError {
+    /// An error occurred when accessing an export
+    Export(ExportError),
+}
+
+impl From<ExportError> for HostEnvInitError {
+    fn from(other: ExportError) -> Self {
+        Self::Export(other)
+    }
+}
+
+/// Prototype trait for finishing envs.
+/// # Examples
+///
+/// This trait can be derived like so:
+///
+/// ```
+/// # use wasmer::{WasmerEnv, LazyInit, Memory, NativeFunc};
+///
+/// #[derive(WasmerEnv)]
+/// pub struct MyEnvWithNoInstanceData {
+///     non_instance_data: u8,
+/// }
+///
+/// #[derive(WasmerEnv)]
+/// pub struct MyEnvWithInstanceData {
+///     non_instance_data: u8,
+///     #[wasmer(export)]
+///     memory: LazyInit<Memory>,
+///     #[wasmer(export(name = "real_name"))]
+///     func: LazyInit<NativeFunc<(i32, i32), i32>>,
+/// }
+///
+/// ```
+///
+/// This trait can also be implemented manually:
+/// ```
+/// # use wasmer::{WasmerEnv, LazyInit, Memory, Instance, HostEnvInitError};
+/// pub struct MyEnv {
+///    memory: LazyInit<Memory>,
+/// }
+///
+/// impl WasmerEnv for MyEnv {
+///     fn finish(&mut self, instance: &Instance) -> Result<(), HostEnvInitError> {
+///         let memory = instance.exports.get_memory("memory").unwrap();
+///         self.memory.initialize(memory.clone());
+///         Ok(())
+///     }
+/// }
+/// ```
+pub trait WasmerEnv {
+    /// The function that Wasmer will call on your type to let it finish
+    /// setting up the environment with data from the `Instance`.
+    ///
+    /// This function is called after `Instance` is created but before it is
+    /// returned to the user via `Instance::new`.
+    fn finish(&mut self, instance: &Instance) -> Result<(), HostEnvInitError>;
```

Comment:
I don't really like the name `finish`. `setup_instance`, `init_from_instance` or other names might be better. Other ideas?

Reply:
Also, why not have a default implementation that returns `Ok(())`?

That will simplify things I think.

Reply:
> I don't really like the name finish. setup_instance, init_from_instance or other names might be better. Other ideas?

Yeah, I agree that it probably needs a better name.

Of those three, the one I like the most is `init_from_instance`.  I'll think about it for a bit 🤔